indico/modules/users/tasks.py

```diff
@@ -37,3 +60,215 @@ def update_gravatars(user=None):
             continue
         set_user_avatar(user, gravatar, 'gravatar', lastmod)
         logger.info('Gravatar of user %s updated', user)
+
+
+def export_user_data(user, options):
+    """Generate a zip file with all user data and save it in user.data_export_request."""
+    export_request = user.data_export_request
+    if export_request and export_request.is_running:
+        return
+
+    # ughh why isn't there an ORM upsert..
+    if export_request:
+        db.session.delete(export_request)
+        db.session.commit()
+    export_request = DataExportRequest(user=user, selected_options=options, state=DataExportRequestState.running)
+    db.session.commit()
+
+    temp_file = NamedTemporaryFile(suffix='.zip', dir=config.TEMP_DIR, delete=False)
+    try:
+        _export_user_data(export_request, temp_file)
+    except Exception:
+        logger.exception('Could not create a user data export %r', export_request)
+        export_request.fail()
+        db.session.commit()
+        notify_data_export_failure(export_request)
+    finally:
+        Path(temp_file.name).unlink(missing_ok=True)
+
+
+def _export_user_data(export_request, buffer):
+    generate_zip(export_request, buffer, config.MAX_DATA_EXPORT_SIZE)
+    buffer.seek(0)
+
+    file = File(filename='data-export.zip', content_type='application/zip')
+    try:
+        file.save(('user', export_request.user.id), buffer)
+        file.claim()
+        export_request.succeed(file)
+        db.session.commit()
+    except Exception as exc:
+        with suppress(Exception):
+            file.delete()
+        raise exc
+    else:
+        notify_data_export_success(export_request)
+
+
+@celery.task(name='export_user_data')
+def export_user_data_task(user, options):
+    export_user_data(user, options)
+
+
+def get_data(export_request):
+    from indico.modules.users.schemas import UserDataExportSchema
+
+    user = export_request.user
+    options = export_request.selected_options
+
+    options_map = {
+        DataExportOptions.contribs: ('contributions', 'subcontributions'),
+        DataExportOptions.abstracts_papers: ('abstracts', 'papers'),
+        DataExportOptions.misc: ('miscellaneous',),
+    }
+    fields = []
+    for opt in options:
+        fields += options_map.get(opt, (opt.name,))
+    return UserDataExportSchema(only=fields).dump(user)
+
+
+def generate_zip(export_request, temp_file, max_size_mb):
+    data = get_data(export_request)
+    with ZipFile(temp_file, 'w', allowZip64=True) as zip_file:
+        zip_file.writestr('/data.yml', yaml.dump(data))
```

Comment:
Hadn't we agreed on creating separate yaml files for each top-level key?

```suggestion
        for key, subdata in data.items():
            zip_file.writestr(f'/{key}.yml', yaml.dump(subdata))
```

Reply:
Have we? I'm fine with either, just genuinely don't remember.. :sweat_smile:

Reply:
at least I thought so... in any case, it makes it much easier to read the files as a human

fwiw i have the suggestion locally since i wanted to test the readability with the separate files so i can push that

Reply:
yeah if you already have it, feel free to push it :)